### ISSUE TITLE
redistributeParams cannot be modified if NeuralNetConfiguration.Builder() is used to construct a configuration

### ISSUE THREAD

**Issue by DennisChen0307** [deeplearning4j/deeplearning4j#839]
I was searching for methods to save and load model of the network and I got  this article:

    http://deeplearning4j.org/modelpersistence.html

The code seemed to be worked for me, but at last I got wrong parameters output from the newly loaded model: all the values are zero. 

Original network params [ 47.32, -47.21, -47.23, -47.27, -47.15, 47.18, 47.34, 47.17, 47.15.....

Saved network params [ 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, ....

Here I use NeuralNetConfiguration.Builder() to build  MultiLayerConfiguration.
I went throught the source code and I believe it is because of this:

public class NeuralNetConfiguration implements Serializable,Cloneable {
    public static class ListBuilder extends MultiLayerConfiguration.Builder {
......
        public MultiLayerConfiguration build() {
            List<NeuralNetConfiguration> list = new ArrayList<>();
            for(int i = 0; i < layerwise.size(); i++) {
                list.add(layerwise.get(i).build());
            }
            return new MultiLayerConfiguration.Builder().backprop(backprop).inputPreProcessors(inputPreProcessors).
                    pretrain(pretrain).backpropType(backpropType).tBPTTForwardLength(tbpttFwdLength)
                    .tBPTTBackwardLength(tbpttBackLength)
                    .redistributeParams(redistributeParams)
                    .confs(list).build();
        }
......

We don't have a interface to modify redistributeParams for a MultiLayerConfiguration in NeuralNetConfiguration so it can only be false by default.

I found there is a very interesting point of this methodology: the model can only be trained once and redistributed many times. But I still believe this is not what a network should really behave. We cannot tell whether it is loaded properly

I paste my code here:

/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.test;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Paths;
import org.apache.commons.io.FileUtils;
import org.deeplearning4j.eval.Evaluation;
import org.deeplearning4j.nn.api.OptimizationAlgorithm;
import org.deeplearning4j.nn.conf.MultiLayerConfiguration;
import org.deeplearning4j.nn.conf.NeuralNetConfiguration;
import org.deeplearning4j.nn.conf.layers.DenseLayer;
import org.deeplearning4j.nn.conf.layers.OutputLayer;
import org.deeplearning4j.nn.multilayer.MultiLayerNetwork;
import org.deeplearning4j.nn.weights.WeightInit;
import org.deeplearning4j.optimize.listeners.ScoreIterationListener;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.dataset.DataSet;
import org.nd4j.linalg.dataset.SplitTestAndTrain;
import org.nd4j.linalg.factory.Nd4j;
import org.nd4j.linalg.lossfunctions.LossFunctions;
import org.springframework.core.io.ClassPathResource;
import weka.core.Instances;
import weka.core.converters.ConverterUtils;
import org.nd4j.linalg.api.ops.impl.transforms.*;

/**
 * Created by agibsonccc on 9/11/14.
 *
 * Diff from small single layer
 */
public class DeepLearner {

    public static void main(String[] args) throws Exception {
        
        /*
        Predictor p = new Predictor(12);

        Instances data1=ConverterUtils.DataSource.read(new ClassPathResource("abc.arff").getPath());

        Instances samples = p.assembleSample(data1);
        
        DataSet ds=Utils.wekaInstanceToDLDataSet(samples);
        */
        
        
        int nSamples = 50;
        INDArray x = Nd4j.linspace(-10, 10, nSamples).reshape(nSamples, 1);
        INDArray y = x.div(2);
        DataSet ds = new DataSet(x, y);
        
        
        ds.normalize();
        
        int seed = 10000;
        int iterations = 100;
        int inputNum = ds.numInputs();
        int hiddenNum = 100;
        double learningRate = 5000;

        MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
                .seed(seed) // Seed to lock in weight initia-lization for tuning
                .iterations(iterations) // # training iterations predict/classify & backprop
                .optimizationAlgo(OptimizationAlgorithm.LINE_GRADIENT_DESCENT)
                .learningRate(learningRate) // Optimization step size
                .constrainGradientToUnitNorm(false)
                
                .list(4) // # NN layers (does not count input layer)
                .layer(0, new DenseLayer.Builder()
                                .nIn(inputNum) // # input nodes
                                .nOut(hiddenNum) // # output nodes
                                .activation("tanh")
                                .weightInit(WeightInit.XAVIER)
                                .build()
                )
                .layer(1, new DenseLayer.Builder()
                                .nIn(hiddenNum) // # input nodes
                                .nOut(hiddenNum*2) // # output nodes
                                .activation("tanh")
                                .weightInit(WeightInit.XAVIER)
                                .build()
                )
                .layer(2, new DenseLayer.Builder()
                                .nIn(hiddenNum*2) // # input nodes
                                .nOut(hiddenNum) // # output nodes
                                .activation("tanh")
                                .weightInit(WeightInit.XAVIER)
                                .build()
                )
                .layer(3, new OutputLayer.Builder(LossFunctions.LossFunction.MSE)
                                .nIn(hiddenNum) // # input nodes
                                .nOut(1) // # output nodes
                                .activation("identity")
                                .weightInit(WeightInit.XAVIER)
                                .build()
                )// NN layer type
                .backprop(true)
                .build();
       
        
        System.out.println(ds.numInputs());
        System.out.println(ds.numExamples());
        System.out.println(ds.getLabels());
        
        
        MultiLayerNetwork network = new MultiLayerNetwork(conf);
        network.init();
        //network.setListeners(new ScoreIterationListener(1));
        network.fit(ds);
        SplitTestAndTrain testAndTrain = ds.splitTestAndTrain(0.9);
        DataSet test = testAndTrain.getTest();
        INDArray testFeature = test.getFeatureMatrix();
        INDArray testLabel = test.getLabels();
        INDArray testPredictedLabel = network.output(testFeature);
        // Fitting model
        // Get predictions from test feature
        /*
        
        
        INDArray yEst = network.output(x);
        INDArray error = Nd4j.std(testPredictedLabel.sub(yEst));

        System.out.println("X" + x);
        System.out.println("Z" + yEst);
        System.out.println("Y" + y);
        System.out.println(error);
        */

        // Eval without class number
        //Evaluation eval2 = new Evaluation(); //// No class num
        
        System.out.println(testLabel);
        System.out.println(testPredictedLabel);
        System.out.println(network.params());
        //eval2.eval(testLabel, testPredictedLabel);
        
        //System.out.println(eval2.stats());
        
        String coef_file = "coefficients.bin";
        String conf_file = "conf.json";
        OutputStream fos = Files.newOutputStream(Paths.get(coef_file));
        DataOutputStream dos = new DataOutputStream(fos);
        Nd4j.write(network.params(), dos);
        dos.flush();
        dos.close();
        FileUtils.write(new File(conf_file), network.getLayerWiseConfigurations().toJson());

        MultiLayerConfiguration confFromJson = MultiLayerConfiguration.fromJson(
                            FileUtils.readFileToString(new File(conf_file)));
        DataInputStream dis = new DataInputStream(new FileInputStream(coef_file));
        INDArray newParams = Nd4j.read(dis);
        System.out.println(newParams);
        dis.close();
        MultiLayerNetwork savedNetwork = new MultiLayerNetwork(confFromJson,newParams);
        savedNetwork.reDistributeParams();
        //savedNetwork.setParameters(newParams);
        //savedNetwork.init();
        System.out.println("Original network params " + network);
        System.out.println("Saved network params " + savedNetwork);
    }

}

**Comment by AlexDBlack** [comment]
Thanks. This is a known issue that has since been fixed: https://github.com/deeplearning4j/deeplearning4j/issues/782

We just pushed the a new version of nd4j and dl4j out today (0.4-rc3.6) that includes the fix (it isn't showing on Maven central just yet, but it's usable already).
